### PR TITLE
resolve #6762 on ldconfig only available on root PATH

### LINKED ISSUE #6762: configure assumes ldconfig can be called and that CuDNN is installed in a system library location
The configure script calls "ldconfig" to locate the CuDNN library but this makes 2 assumptions that can be wrong: 
1. It assumes that "ldconfig" can be called. However, on some GNU/Linux systems such as openSUSE Leap 42.1, only root has it in their path. (On Leap 42.1, normal users get the desired output from "/sbin/ldconfig -p".)
2. It assumes that the CuDNN library is installed in a system library location that is searched by ldconfig. However, users may install it in their home directory and use LD_LIBRARY_PATH to load it for various reasons. I, for example, was not fully satisfied from reading the licence terms that I am allowed to share the library with other users of the system and therefore did not install it system-wide as root.

Severity is low as normally the CuDNN library location should be auto-detected from the CuDNN library path before "ldconfig" is tried.

How to reproduce:
1. git clone this repo on a system with ldconfig not in your PATH
2. cd tensorflow and run ./configure
3. when asked for the CuDNN path, enter a valid and readable path (to avoid other errors), for example /usr/bin, that does not contain CuDNN
4. Error message: ./configure: line 337: ldconfig: command not found

Workaround options:
- First double check the CuDNN library path (the default path suggests that the last component "lib64" or "lib" should be omitted) and verify that standard folder names are used under this location for library and header files, for example "lib64" and "include".
- If you can call /sbin/ldconfig and if CuDNN is in a standard system library location the user can temporarily add /sbin to their PATH before running "./configure"
- Edit the "configure" script to provide the location manually, for example CUDNN_PATH_FROM_LDCONFIG="$HOME/local/cuda/lib64/libcudnn.so"

A fix should test whether "ldconfig" can be called before trying to call it. Also try /sbin/ldocnfig. Given that the library should normally be found already earlier in the code using the user-provided CuDNN path, a simple solution may be to set CUDNN_PATH_FROM_LDCONFIG to the empty string if ldconfig cannot be called.

If you like to automate the locating of the library as much as possible, you could furthermore check all locations in LD_LIBRARY_PATH.

### PR DESCRIPTION
**Pull request by byronyi**
See Linux man page for [ldconfig(8)](https://linux.die.net/man/8/ldconfig), synopsis section:

```bash
/sbin/ldconfig [ -nNvXV ] [ -f conf ] [ -C cache ] [ -r root ] directory ...
/sbin/ldconfig -l [ -v ] library ...
/sbin/ldconfig -p
```

Former pull request #6848 has been closed as @jowagner has not signed a CLA.

### CODE REVIEW

**Comment by jowagner** [comment]
This patch may break on some platforms. The FHS states that ldconfig is "usually located in /usr/sbin". The alternative location "/sbin" is described as optional.

**Comment by jowagner** [comment]
openSUSE and Debian: ok (/sbin)

**Comment by byronyi** [comment]
Can someone review this PR? Current master with CUDA still breaks on latest Debian stable.

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

If this passes all of our tested / supported platforms, it seems okay, but will let Gunhan review when he is free (he is out of the office for a bit though).  Or maybe @caisq  if he feels trongly

File: configure [hunk 1]
```diff
@@ -456,7 +456,7 @@ while true; do
   fi
 
   if is_linux; then
-    CUDNN_PATH_FROM_LDCONFIG="$(ldconfig -p | sed -n 's/.*libcudnn.so .* => \(.*\)/\1/p')"
+    CUDNN_PATH_FROM_LDCONFIG="$(/sbin/ldconfig -p | sed -n 's/.*libcudnn.so .* => \(.*\)/\1/p')"
```
Comment (gunan) [inline configure]: I feel like we need to be compatible with the previous case, where ldconfig is not under `/sbin`, but is in `PATH`.
For example, I have a centos system which is happy the way things are right now, but will break with this change as ldconfig resides in `/usr/sbin/ldconfig`
Can we modify these lines to say:
```
if ! type ldconfig > /dev/null; then
  LDCONFIG_BIN=/sbin/ldconfig
else
  LDCONFIG_BIN=ldconfig
fi
CUDNN_PATH_FROM_LDCONFIG="$($LDCONFIG_BIN -p | ......
```

**Comment by byronyi** [comment]
@gunan Thanks for your help. I've changed the lines as suggested.

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by jowagner** [comment]
An error message "configure: line xyz: type: ldconfig: not found" is printed on stderr if the command is not found in the line starting with "if ! type ldconfig".

There's a simple solution but as I cannot sign the CLA at this time, I cannot submit it. Hint: 5 characters.

I can confirm, however, that the variable LDCONFIG_BIN is initialised as intended to "/sbin/ldconfig" for my login without ldconfig in PATH.

Tested on openSUSE Leap 42.2.

**Comment by byronyi** [comment]
@jowagner Thanks for the heads up. Not sure why some checks failed though.

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by vrv** [comment]
gpu build passed, so I'm calling this a win.

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by byronyi** [comment]
That's my first PR merged into TF! Thank you guys a lot @vrv @gunan @jowagner for your guide and suggestion. 

Really remarkable moment to me and I hope I could submit more patches in the future.

**Comment by jowagner** [comment]
Thank you for incrementally coming up with a solution that is substantially different to my CLA-less PR and for taking the time to bring this over the finishing line.

**Pull request merged**